### PR TITLE
Add the importer

### PR DESCRIPTION
**Pull request by thomas-hilaire**

### CODE REVIEW

File: src/importer.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    constructor(
+        private papiClient: PapiClient,
+        private config: ImporterConfig,
+        private amqpConnectionProvider
+    ) {}
+
+    public importAllEvents(): void {
+        let runCount = 0;
+        this.amqpConnectionProvider
+            .flatMap(connection => connection.createChannel())
```
Comment (mbaechler) [inline src/importer.ts]: why flatmap and not map ? it looks like you only transform the value, you don't generate another stream of values

File: src/importer.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            .flatMap(channel => channel.assertQueue(this.config.amqp.queues.event, { durable: true }))
+            .flatMap(reply => {
+                reply.channel.prefetch(this.config.maxBatchSize);
+                return reply.channel.consume(this.config.amqp.queues.event, { noAck: false });
+            })
+            .bufferWithTimeOrCount(this.config.maxBatchWaitTimeMs, this.config.maxBatchSize)
+            .subscribe(events => {
```
Comment (mbaechler) [inline src/importer.ts]: you could extract the lambda in subscribe to name it

File: src/importer.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        console.log("Starting a batch");
+        return this.papiClient.startBatch().then(() => {
+            return this.papiClient.importAllICS(this.messagesToPapiEvents(events))
+                .then(() => {
+                    console.info("Commiting a batch");
+                    return this.papiClient.commitBatch();
+                })
+                .then(() => {
+                    console.info("Waiting for batch to finish");
+                    return this.papiClient.waitForBatchSuccess();
```
Comment (mbaechler) [inline src/importer.ts]: why do you want to wait ? aren't you responsive ?

**Comment by ddolcimascolo** [comment]
:+1:

File: src/importer.ts (hunk 1 above)
Comment (thomas-hilaire) [inline src/importer.ts]: I agree you however it's how the rx-amqplib is built (see https://github.com/SkippyZA/rx-amqplib). I did not investigate about that, but using `map` does not work there.

File: src/importer.ts (hunk 2 above)
Comment (thomas-hilaire) [inline src/importer.ts]: done

File: src/importer.ts (hunk 1 above)
Comment (fvn-linagora) [inline src/importer.ts]: That's how Observable works actually. createChannel returns an observable "stream" (onNext, .. onNext, onComplete) even for single observables (onNext and onComplete only).
Hence using flatMap is used to chain observables items (onComplete->onNext), as a way to serialize/sequence events.

**Pull request merged**